addons/t9n/models/message.py

```
@@ -11,14 +11,67 @@ class Message(models.Model):
+    )
     resource_id = fields.Many2one(
         comodel_name="t9n.resource",
         help="The resource (typically a file) from which the entry is coming from.",
+        ondelete="cascade",
+        required=True,
     )
     translation_ids = fields.One2many(
         comodel_name="t9n.translation",
         inverse_name="source_id",
         string="Translations",
     )
+
+    _sql_constraints = [
+        (
+            "body_context_resource_unique",
+            "UNIQUE(body, context, resource_id)",
+            "The combination of the message body and context must be unique within the same resource!"
+        ),
+    ]
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        updated_records = self.env["t9n.message"]
+
+        resource_ids = list(set(vals.get("resource_id") for vals in vals_list if "resource_id" in vals))
+        resources = self.env["t9n.resource"].browse(resource_ids)
+        project_resources_map = {res.id: res.project_id for res in resources}
+
+        bodies = set(vals.get("body") for vals in vals_list)
+        contexts = set(vals.get("context") for vals in vals_list)
+
+        potential_matches = self.search([
+            ("body", "in", list(bodies)),
+            ("context", "in", list(contexts))
+        ])
+
+        for vals in vals_list:
+            cur_project_id = project_resources_map.get(vals.get("resource_id"))
+
+            matching_records = potential_matches.filtered(lambda r: r["body"] == vals.get(
+                "body") and ((r["context"] == vals.get("context")) or (not r["context"] and vals.get("context") is None)) and r["resource_id"].project_id == cur_project_id)
+
+            if matching_records:
+                existing_record = matching_records[0]
+                existing_record.write(vals)
+                updated_records += existing_record
+            else:
+                updated_records += super().create([vals])
+        return updated_records
```

It feels weird to me to do some filtering inside a create method. If I give it a set of values, I expect it to create a record for each of those values. If I gave it crap, I'd expect the code, e.g. the SQL constraint, to yell at me instead of silently ignoring the bad data.

So I'd move the filter logic higher up in the flow if possible, not right inside the create method